osquery/tables/system/darwin/crash_logs.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include "osquery/tables/system/system_utils.h"
+#include <osquery/core/conversions.h>
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/trim.hpp>
+#include <boost/format.hpp>
+
+namespace fs = boost::filesystem;
+namespace alg = boost::algorithm;
+
+namespace osquery {
+namespace tables {
+
+const std::map<std::string, std::string> crashDumpKeys = {
+    {"Process", "process"},
+    {"Path", "application_path"},
+    {"Identifier", "identifier"},
+    {"Version", "version"},
+    {"Parent Process", "parent_proc"},
+    {"Responsible", "responsible"},
+    {"User ID", "user_id"},
+    {"Date/Time", "date_time"},
+    {"Crashed Thread", "crashed_thread"},
+    {"Exception Type", "exception_type"},
+    {"Exception Codes", "exception_codes"},
+    {"Exception Note", "exception_notes"},
+    {"rax", "rax"},
+    {"rbx", "rbx"},
+    {"rcx", "rcx"},
+    {"rdx", "rdx"},
+    {"rdi", "rdi"},
+    {"rsi", "rsi"},
+    {"rbp", "rbp"},
+    {"rsp", "rsp"},
+};
+
+void readCrashDump(const std::string& app_log, Row& r) {
+  auto apl = fs::path(app_log);
+  r["log_location"] = app_log;
+  std::string content;
+
+  if (readFile(apl, content).ok()) {
+
+    // Variables for capturing the stack trace
+    boost::format crashed_thread_format("Thread %1% Crashed");
+    auto crashed_thread_seen = false;
+
+    auto lines = split(content, "\n");
+    for (auto it = lines.begin(); it != lines.end(); it++) {
+      auto line = *it;
+      // Tokenize each line by a colon, to get key/value pairs.
+      auto toks = split(line, ":");
+
+      // Grab the most recent stack trace line of the crashed thread.  If we so
+      // choose, we could grab the rest of the stack context here, or at least
+      // more frames.
+      if (crashed_thread_seen && toks[0] == crashed_thread_format.str()) {
```

Comment:
What happens if `line.size() == 0`? Will `toks.size() == 0`? If so then this is an `abrt()`.